File: android/app/src/main/java/binding/StageBinding.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2023 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package binding

import channel.command.CommandName
import channel.stage.StageModal
import channel.stage.StageOps
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.launch
import model.Tab
import org.blokada.R
import service.AlertDialogService
import service.ContextService
import service.FlutterService
import service.Sheet
import service.SheetService
import ui.utils.openInBrowser

object StageBinding: StageOps {
    private val writeForeground = MutableStateFlow<Boolean?>(null)
    val enteredForegroundHot = writeForeground.filterNotNull().filter { it }

    val route = MutableStateFlow("home")
    val tab = MutableStateFlow(Tab.Home)
    val payload = MutableStateFlow("")

    private val context by lazy { ContextService }
    private val flutter by lazy { FlutterService }
    private val command by lazy { CommandBinding }
    private val sheet by lazy { SheetService }
    private val dialog by lazy { AlertDialogService }
    private val scope = GlobalScope

    var showingNavBar = true
    var onShowNavBar: (Boolean) -> Unit = { }
    private var displayingModal: StageModal? = null

    private var previousRoute = ""

    init {
        StageOps.setUp(flutter.engine.dartExecutor.binaryMessenger, this)
    }

    fun setRoute(route: String) {
        previousRoute = this.route.value
        scope.launch {
            command.execute(CommandName.ROUTE, route)
        }
    }

    fun goBack(): Boolean {
        if (route.value.startsWith("http")) {
            setRoute(previousRoute)
            return true
        } else if (route.value.contains("/")) {
            setRoute(route.value.substringBeforeLast("/"))
            return true
        } else {
            return false
        }
    }

    fun setForeground() {
        scope.launch {
            command.execute(CommandName.FOREGROUND)

            // Special case handling: android wont give us a call when the share sheet is dismissed
            if (displayingModal == StageModal.ADSCOUNTERSHARE) {
                modalDismissed()
            }
        }
    }

    fun setBackground() {
        scope.launch {
            command.execute(CommandName.BACKGROUND)
        }
    }

    fun dismiss() {
        scope.launch {
            command.execute(CommandName.MODALDISMISS)
        }
    }

    fun sheetShown(sheet: Sheet) {
        val name = when (sheet) {
            Sheet.Payment -> StageModal.PAYMENT
            Sheet.Activated -> StageModal.PERMS
            Sheet.Location -> StageModal.PLUSLOCATIONSELECT
            Sheet.Help -> StageModal.HELP
            Sheet.Custom -> StageModal.CUSTOM
            Sheet.AccountChange -> StageModal.ACCOUNTCHANGE
            else -> null
        }

        if (name != null) {
            scope.launch {
                command.execute(CommandName.MODALSHOWN, name.name)
            }
        }
    }

    fun modalDismissed() {
        displayingModal = null
        scope.launch {
            command.execute(CommandName.MODALDISMISSED)
        }
    }

    override fun doShowModal(modal: StageModal, callback: (Result<Unit>) -> Unit) {
        if (displayingModal == modal) return callback(Result.success(Unit))
        if (showDialog(modal)) {
            displayingModal = modal
            return callback(Result.success(Unit))
        }

        val name = when (modal) {
            StageModal.PAYMENT -> Sheet.Payment
            StageModal.PERMS -> Sheet.Activated
            StageModal.PLUSLOCATIONSELECT -> Sheet.Location
            StageModal.HELP -> Sheet.Help
            StageModal.CUSTOM -> Sheet.Custom
            StageModal.ACCOUNTCHANGE -> Sheet.AccountChange
            else -> null
        }

        if (name != null) {
            sheet.showSheet(name)
        } else {
            // Confirm modal was shown for unknown modals, so the common lib can display it
            scope.launch {
                command.execute(CommandName.MODALSHOWN, modal.name)
            }
        }

        displayingModal = modal
        callback(Result.success(Unit))
    }

    private fun showDialog(modal: StageModal): Boolean {
        val name = when (modal) {
            StageModal.FAULT, StageModal.ACCOUNTINITFAILED -> R.string.error_unknown
            StageModal.ACCOUNTEXPIRED -> R.string.error_account_inactive_generic
            StageModal.ACCOUNTRESTOREFAILED -> R.string.error_payment_inactive_after_restore
            StageModal.ACCOUNTRESTOREIDFAILED -> R.string.error_account_inactive_after_restore
            StageModal.ACCOUNTRESTOREIDOK -> R.string.generic_account_active
            StageModal.ACCOUNTINVALID -> R.string.error_account_invalid
            StageModal.PLUSTOOMANYLEASES -> R.string.error_vpn_too_many_leases
            StageModal.PLUSVPNFAILURE -> R.string.error_vpn
            StageModal.PAYMENTUNAVAILABLE -> R.string.error_payment_not_available
            StageModal.PAYMENTTEMPUNAVAILABLE -> R.string.error_payment_failed
            StageModal.PAYMENTFAILED -> R.string.error_payment_failed_alternative
            else -> null
        }

        val title = when (modal) {
            StageModal.ACCOUNTRESTOREIDOK -> R.string.payment_header_activated
            else -> null
        }

        val action: Pair<String, () -> Unit>? = when (modal) {
            StageModal.ACCOUNTRESTOREIDOK -> null
            else -> {
                context.requireContext().getString(R.string.universal_action_show_log) to {
                    GlobalScope.launch {
                        command.execute(CommandName.LOG)
                    }
                }
            }
        }

        if (name != null) {
            dialog.showAlert(name, title = title, onDismiss = ::modalDismissed, additionalAction = action)
            scope.launch {
                command.execute(CommandName.MODALSHOWN, modal.name)
            }
            return true
        }
        return false
    }

    override fun doDismissModal(callback: (Result<Unit>) -> Unit) {
        if (displayingModal == StageModal.ADSCOUNTERSHARE) {
            modalDismissed()
        } else {
            displayingModal = null
            sheet.dismiss()
            dialog.dismiss()
        }
        callback(Result.success(Unit))
    }

    override fun doOpenLink(url: String, callback: (Result<Unit>) -> Unit) {
        openInBrowser(url)
        callback(Result.success(Unit))
    }

    override fun doHomeReached(callback: (Result<Unit>) -> Unit) {
        context.requireActivity().finish();
        callback(Result.success(Unit))
    }
}

File: ios/App/Binding/Stage/StageBinding.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2023 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Factory
import Combine
import UIKit

extension StageModal: Identifiable {
    var id: Int {
        rawValue
    }
}

class StageBinding: StageOps {

    let enteredForegroundHot = CurrentValueSubject<Bool?, Never>(nil)
    let currentModal = CurrentValueSubject<StageModal?, Never>(nil)
    let showPauseMenu = CurrentValueSubject<Bool, Never>(false)
    let errorHeader = CurrentValueSubject<String?, Never>(nil)
    let error = CurrentValueSubject<Error?, Never>(nil)

    let showInput = CurrentValueSubject<Bool, Never>(false)
    
    let netx = Services.netx

    @Injected(\.flutter) private var flutter
    @Injected(\.commands) private var commands

    func setTab(_ tab: Tab) {
        commands.execute(.route, "\(tab.rawValue.lowercased())")
    }

    func setRoute(_ path: String) {
        commands.execute(.route, path)
    }

    func onForeground(_ foreground: Bool) {
        enteredForegroundHot.send(foreground)
        if foreground {
            commands.execute(.foreground)
            netx.refreshOnForeground() // TODO: a better place
        } else {
            commands.execute(.background)
        }
    }
    
    func showModal(_ modal: StageModal, params: Any? = nil) {
        commands.execute(.modalShow, "\(modal)")
    }

    func dismiss() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            self.commands.cmd.onCommand(command: "\(CommandName.modalDismiss)", m: 2) { _ in
                return promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }

    func onDismissed() {
        commands.execute(.modalDismissed)
    }

    func showPauseMenu(_ show: Bool) {
        if (show) {
            commands.execute(.modalShow, "pause")
        } else {
            commands.execute(.modalDismissed)
        }
    }

    init() {
        StageOpsSetup.setUp(binaryMessenger: flutter.getMessenger(), api: self)
    }

    var supportedSheets: [StageModal] = [
        .plusLocationSelect, .accountChange
    ]

    func doShowModal(modal: StageModal, completion: @escaping (Result<Void, Error>) -> Void) {
//    if (modal == .pause) {
//        currentModal.send(nil)
//        showPauseMenu.send(true)
//    } else

        errorHeader.send(nil)

        if supportedSheets.contains(modal) {
            currentModal.send(modal)
        } else if (modal == .fault) {
            error.send(L10n.errorUnknown)
        } else if (modal == .faultLocked) {
            error.send(L10n.errorLocked)
        } else if (modal == .faultLockInvalid) {
            //error.send(L10n.errorLockInvalid)
        } else if (modal == .faultLinkAlready) {
            error.send(L10n.familyFaultLinkAlready)
        } else if (modal == .accountInitFailed) {
            error.send(L10n.errorUnknown)
        } else if (modal == .accountRestoreFailed) {
            error.send(L10n.errorPaymentInactiveAfterRestore)
        } else if (modal == .accountRestoreIdOk) {
            errorHeader.send(L10n.paymentHeaderActivated)
            error.send(L10n.genericAccountActive)
        } else if (modal == .accountRestoreIdFailed) {
            error.send(L10n.errorAccountInactiveAfterRestore)
        } else if (modal == .accountExpired) {
            //error.send(L10n.errorAccountInactiveGeneric)
        } else if (modal == .plusTooManyLeases) {
            error.send(L10n.errorVpnTooManyLeases)
        } else if (modal == .plusVpnFailure) {
            error.send(L10n.errorVpn)
        } else if (modal == .paymentUnavailable) {
            error.send(L10n.errorPaymentNotAvailable)
        } else if (modal == .paymentTempUnavailable) {
            error.send(L10n.errorPaymentFailed)
        } else if (modal == .paymentFailed) {
            error.send(L10n.errorPaymentFailedAlternative)
        } else if (modal == .accountInvalid) {
            error.send(L10n.errorAccountInvalid)
        } else if (modal == .deviceAlias) {
            showInput.send(true)
        } else {
            currentModal.send(nil)
        }
        completion(.success(()))
        commands.execute(.modalShown, "\(modal)")
    }
    
    func doDismissModal(completion: @escaping (Result<Void, Error>) -> Void) {
        if currentModal.value != nil {
            currentModal.send(nil)
            //showPauseMenu.send(false)
            completion(.success(()))
            commands.execute(.modalDismissed)
        } else if error.value != nil {
            error.send(nil)
            completion(.success(()))
            commands.execute(.modalDismissed)
        } else if showInput.value {
            showInput.send(false)
            completion(.success(()))
            commands.execute(.modalDismissed)
        } else {
            //showPauseMenu.send(false)
            completion(.success(()))
            commands.execute(.modalDismissed)
        }
    }

    func doOpenLink(url: String, completion: @escaping (Result<Void, Error>) -> Void) {
        if let link = URL(string: url) {
          UIApplication.shared.open(link)
        }
        completion(.success(()))
    }

    func doHomeReached(completion: @escaping (Result<Void, any Error>) -> Void) {
        // not used on ios
        completion(.success(()))
    }
}

extension Container {
    var stage: Factory<StageBinding> {
        self { StageBinding() }.singleton
    }
}
